python/lsst/daf/persistence/policy.py

```diff
@@ -0,0 +1,392 @@
+
+        :param other:
+        :return:
+        """
+        other.update(self)
+        self.data = other.data
+
+    def getSubPolicyVal(self, subPolicyName, subPolicyKey):
+        """Look for a subpolicy and if found, return the value for a key.
+
+        :param subPolicyName: the name of a subpolicy to look for. Can be shorter than the full name.
+        E.g. 'raw' may resolve to 'exposures.raw'. If more than one possibility is found, will raise.
+        :param subPolicyKey: the name of a key wihin subpolicy whose value will be returned.
+        :return: the value for subpolicy and key if found, else None.
+        """
+        if subPolicyName is None:
+            return None
+
+        # the ref name can be nested in a hierarchy, e.g. 'raw' could be in 'exposures.raw',
+        # so get the fully resolved name:
+        fullRefPolicyName = [n for n in self.names() if n.endswith(subPolicyName)]
+
+        # make sure we found only 1,
+        # and remove it from the list that the above loop created:
+        if len(fullRefPolicyName) > 1:
+            raise RuntimeError("For policy:%s, found more than 1 possibility for full name:%s"
+                               %(subPolicyName, fullRefPolicyName))
+        elif len(fullRefPolicyName) == 0:
+            return None
+
+        if subPolicyKey in self[fullRefPolicyName[0]]:
+            return self[fullRefPolicyName[0]][subPolicyKey]
+
+        return None
+
+    def names(self, topLevelOnly=False):
+        """Get the dot-delimited name of all the keys in the hierarchy.
+        NOTE: this is different than the built-in method dict.keys, which will return only the first level
+        keys.
+        """
+        if topLevelOnly:
+            return self.keys()
+
+        def getKeys(d, keys, base):
+            for key in d.keys():
+                val = d[key]
+                levelKey = base + '.' + key if base is not None else key
+                keys.append(levelKey)
+                if isinstance(val, collections.Mapping):
+                    getKeys(val, keys, levelKey)
+        keys = []
+        getKeys(self.data, keys, None)
+        return keys
+
+    def asArray(self, name):
+        """Get a value as an array. May contain one or more elements.
+
+        :param key:
+        :return:
+        """
+        val = self.get(name)
+        if isinstance(val, basestring):
+            val = [val]
+        elif not isinstance(val, collections.Container):
+            val = [val]
+        return val
+
+    # Deprecated methods that mimic pex_policy api.
+    # These are supported (for now), but callers should use the dict api.
+
+    def getValue(self, name):
+        """Get the value for a parameter name/key. See class notes about dot-delimited access.
+
+        :param name:
+        :return: the value for the given name.
+        """
+        warnings.warn_explicit("Deprecated. Use []", DeprecationWarning)
```

Comment:
if there's a better way to mark methods as deprecated I'd like to know about it. These don't actually do anything; the DeprecationWarning is ignored by default. But it's more than nothing at least.